### PR TITLE
[6571] Windmill service flag

### PR DESCRIPTION
**Pull request by dustin12**
Adds the flag --enableStreamingEngine to Java and Python.  Setting this flag turns on the correct experiments to enable streaming engine on the back end.
------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

File: sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/options/GcpOptions.java [hunk 1]
```diff
@@ -126,6 +127,19 @@
 
   void setGcpCredential(Credentials value);
 
+  /** Experiment to turn on the Streaming Engine experiment. */
+  String STREAMING_ENGINE_EXPERIMENT = "enable_streaming_engine";
```
[the remark this answers is not in the document]
Comment (dustin12) [reply sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/options/GcpOptions.java]: Already in an interface.

File: sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/options/GcpOptions.java [hunk 2]
```diff
@@ -126,6 +127,19 @@
...
+
+  /** @deprecated Use STREAMING_ENGINE_EXPERIMENT instead. */
+  @Deprecated String WINDMILL_SERVICE_EXPERIMENT = "enable_windmill_service";
```
[the remark this answers is not in the document]
Comment (dustin12) [reply sdks/java/extensions/google-cloud-platform-core/src/main/java/org/apache/beam/sdk/extensions/gcp/options/GcpOptions.java]: same

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java [hunk 3]
```diff
@@ -318,6 +318,22 @@ public Job translate(List<DataflowPackage> packages) {
...
+          experiments = new ArrayList<String>();
+        }
+        if (!experiments.contains("enable_windmill_service")) {
+          experiments.add("enable_windmill_service");
+        }
+        if (!experiments.contains("enable_streaming_engine")) {
+          experiments.add("enable_streaming_engine");
+        }
+        options.setExperiments(experiments);
+      }
```
[the remark this answers is not in the document]
Comment (dustin12) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java]: Done

**Comment by dustin12** [comment]
Run Java PreCommit

**Comment by dustin12** [comment]
Run Java PreCommit

**Comment by dustin12** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by dustin12** [comment]
Run Spotless PreCommit

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java [hunk 4]
```diff
@@ -318,6 +319,32 @@ public Job translate(List<DataflowPackage> packages) {
...
+          experiments.add(GcpOptions.WINDMILL_SERVICE_EXPERIMENT);
+        }
+        options.setExperiments(experiments);
+      } else {
+        List<String> experiments = options.getExperiments();
+        if (experiments != null) {
+          if (experiments.contains(GcpOptions.STREAMING_ENGINE_EXPERIMENT)
+              || experiments.contains(GcpOptions.WINDMILL_SERVICE_EXPERIMENT)) {
+            throw new IllegalArgumentException(
+                "Streaming engine both disabled and enabled.  Please use --enableStreamingEngine.");
```
Comment (drieber) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java]: I think a better error message would help avoid confused customers:
How about: 
"Streaming engine both disabled and enabled: enableStreamingEngine is set to false, but enable_windmill_service and/or enable_streaming_engine are present. It is recommended you only set enableStreamingEngine."

Comment (dustin12) [reply runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java]: Fixed

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 5]
```diff
@@ -361,6 +362,15 @@ def run_pipeline(self, pipeline, options):
...
       debug_options.experiments = experiments
 
+    # Elevate "enable_streaming_engine" to pipeline option, but using the
+    # existing experiment.
+    google_cloud_options = options.view_as(GoogleCloudOptions)
+    if google_cloud_options.enable_streaming_engine:
+      experiments = ["enable_windmill_service", "enable_streaming_engine"]
+      if debug_options.experiments is not None:
+        experiments = list(set(experiments + debug_options.experiments))
+      debug_options.experiments = experiments
```
[the remark this answers is not in the document]
Comment (dustin12) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: done

**Comment by dustin12** [comment]
Run Spotless PreCommit

**Comment by dustin12** [comment]
Run Java PreCommit

**Comment by dustin12** [comment]
Run Python PreCommit

File: sdks/python/apache_beam/options/pipeline_options.py [hunk 6]
```diff
@@ -410,6 +410,9 @@ def _add_argparse_args(cls, parser):
                         'Experimental. '
                         'See https://cloud.google.com/dataflow/pipelines/'
                         'updating-a-pipeline')
+    parser.add_argument('--enable_streaming_engine',
```
Comment (aaltay) [inline sdks/python/apache_beam/options/pipeline_options.py]: Do you need action=store_true (e.g. https://github.com/apache/beam/blob/master/sdks/python/apache_beam/options/pipeline_options.py#L300) ?

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 7]
```diff
@@ -361,6 +362,27 @@ def run_pipeline(self, pipeline, options):
...
+      if "enable_windmill_service" not in debug_options.experiments:
+        experiments.append("enable_windmill_service")
+      if "enable_streaming_engine" not in debug_options.experiments:
+        experiments.append("enable_streaming_engine")
+      if debug_options.experiments is not None:
+        experiments = list(set(experiments + debug_options.experiments))
+      debug_options.experiments = experiments
+    else:
+      if debug_options.experiments is not None:
+        if ("enable_windmill_service" in debug_options.experiments
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: You are checking twice for ""enable_windmill_service" in debug_options.experiments". Did you intend to check for another experiment?

File: sdks/python/apache_beam/options/pipeline_options.py (hunk 6 above)
Comment (dustin12) [reply sdks/python/apache_beam/options/pipeline_options.py]: added

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py (hunk 7 above)
Comment (dustin12) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: fixed, thanks

**Comment by dustin12** [comment]
Running a sample job gives (abrieviated):

> windmill_service_environment_filter: experiment enable_windmill_service is unsupported on the specified SDK version

which is correct for now as "Windmill Service requires (Dataflow or Beam) Java SDK >= 2.8.0".  It still seems good to keep the flags in sync though so leaving the python flag in for now.

**Pull request merged**